File: backend/app/data_contracts/canonical_affix.py
"""Canonical affix contract."""

from __future__ import annotations

from dataclasses import dataclass, field

from .canonical_base import CanonicalRecord
from .canonical_modifier import CanonicalModifierReference


@dataclass(frozen=True)
class CanonicalAffix(CanonicalRecord):
    affix_type: str | None = None
    tier_count: int | None = None
    item_applicability: tuple[str, ...] = field(default_factory=tuple)
    class_restrictions: tuple[str, ...] = field(default_factory=tuple)
    modifier_references: tuple[CanonicalModifierReference, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "item_applicability", tuple(self.item_applicability))
        object.__setattr__(self, "class_restrictions", tuple(self.class_restrictions))
        object.__setattr__(self, "modifier_references", tuple(self.modifier_references))

File: frontend/src/types/canonicalAffix.ts
import type { CanonicalRecord } from "./canonicalBase";
import type { CanonicalModifierReference } from "./canonicalModifier";

export interface CanonicalAffixTierRange {
  tier: number | string | null;
  min_value: number | null;
  max_value: number | null;
  tier_group?: string | null;
  value_scale?: string | null;
  polarity?: "positive" | "negative" | "mixed" | "unknown" | string;
}

export interface CanonicalAffix extends CanonicalRecord {
  affix_type?: string | null;
  prefix_suffix?: string | null;
  source_type?: string | null;
  affix_domain: "equipment" | "idol" | "unknown";
  tier_count?: number | null;
  item_applicability: string[];
  slot_restrictions: string[];
  item_type_restrictions: string[];
  class_restrictions: string[];
  mastery_restrictions: string[];
  tier_ranges: CanonicalAffixTierRange[];
  modifier_references: CanonicalModifierReference[];
  modifier_reference_count?: number | null;
  value_scale_policy: string;
  polarity_policy: string;
  stable_calculable: boolean;
}
